### ISSUE TITLE
Carrying Over Favorites From 1.x

### ISSUE THREAD

**Issue by akaplo** [umts/pvta-multiplatform#369]
The web version of PVTrAck does this without issue.

The mobile version, however, is another story.  The way we stored favorites/etc in 1.x was restricted to using the webview's implementation of localstorage.

In Ionic 2 the docs tell us this is no longer the case for the following reason:
> When running in a native app context, Storage will prioritize using SQLite, as it's one of the most stable and widely used file-based databases, and avoids some of the pitfalls of things like localstorage and IndexedDB, such as the OS deciding to clear out such data in low disk-space situations.

We can override this decision by telling the framework it **must** use localstorage and therefore enable our upgrading users to keep their favorites, but we run into the pitfalls described above.  This is a 60/40 tradeoff, really.  Either our users keep their favorites now and run the risk of losing them in the future, or they definitely lose them now but are significantly less likely to lose them in the future.

I'm interested in thoughts from @werebus, @sherson , and @petermarathas.  I won't ship 2.0 to the Play Store until a consensus is reached.

**Comment by werebus** [comment]
I'm curious to know how much of on ordeal some sort of... migration tool ❔ would be.  That is can we write something that manually fishes around in localstorage for favorites, adds them to the default app storage, then deletes the localstorage store? After a time, we could remove that feature (when analytics show us that most people have upgraded, perhaps?)

If that's an impractical solution, then I'm okay with switching now. I *think* that users will be slightly more understanding if the loss of their favorites happens at the same time as an obvious (because of the UI changes) upgrade, rather than if the app just randomly "forgets" their favorites occasionally.

**Comment by akaplo** [comment]
That would require us to be able to change the storage driver at runtime, which I [don't think we can do](https://github.com/driftyco/ionic-storage/blob/master/src/storage.ts).  localforage, which IonicStorage is simply a wrapper for, can do this with its `setInstance()` method, so I'm unsure why Ionic didn't expose that.

[This forum answer](https://forum.ionicframework.com/t/multiple-instances-support-in-ionic-storage-v-1-1-7/73875) sums it up nicely ("no, just use localforage at that point").

**Comment by werebus** [comment]
Yeah, you definitely don't want to do multiple IonicStorage instances.  But, surely the [lower-level access](https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage)  is there too, right.

**Comment by sherson** [comment]
Is there an easy (and non-annoying) way to notify users with something to the effect of:

> Unfortunately, there wasn't a reliable way for us to retain a user's saved favorites while upgrading from version 1 to version 2. However, this shouldn't be an issue with future updates. We apologize for the inconvenience.

(also, just to confirm that I didn't misunderstand... is that actually true?)

**Comment by akaplo** [comment]
@sherson you understand correctly, and @werebus you make a great point.  This is feasible using the method described by the accepted answer [here](https://stackoverflow.com/questions/8419354/get-html5-localstorage-keys), combined with the fact that we have a `returningUser` property stored in the original storage.

I'll work something out now and get back to you.

**Comment by akaplo** [comment]
@sherson you understand correctly, and @werebus you make a good point, but there's no guarantee of what type of storage was being used on each user's device (local*Forage*, from 1.x, tries to use IndexedDB, WebSql, and local*Storage* in that order).  

We have a few options available:

1. Turns out removing the CordovaSQLite plugin will cause IonicStorage (2.x) to revert to the same behavior local*Forage* uses.  Easily keep old favorites, possibly OS will remove them if low on space.  Tested this, can confirm it works.

2. Require local*Forage* in 2.x; use it to access the database from 1.x and pull it into 2.x.  Keeps old favorites, OS won't remove them if low on storage, but requires more work and adding local*Forage* to the app (296KB 👻 )

3. Do nothing.  Lose old favorites on native Android devices only, but requires no work.

~~I recommend option 1.  Honestly, if we did it without a second thought in 1.x, why change?~~
